### DISCUSSION TITLE
Creating a NESTED canvas?

### DISCUSSION THREAD

**Discussion by hchaudhary1** [xyflow/xyflow#3945] [Q&A] [answered]
**UseCase**
I want to implement something like this:
https://nocode-js.com/examples/sequential-workflow-designer-pro/webpack-pro-app/public/folders.html
Where the user can drop on a "FOLDER" icon, and it can open a new sub-canvas.

**Questions**
1. Is this possible with the current framework?
2. How much customization would be needed? (i.e. would this be possible by a beginner?)

**Context**
I have a massive canvas with tons of nodes... which slows down when opened, instead if I create sub-groups, the user can open the small sub groups at a time.

**Comment by ozgurrgul** [comment]
> Is this possible with the current framework?

Yes. Once user click on a node, you can add a new node to that one, set the parentNode as the current node's id. 

> How much customization would be needed? (i.e. would this be possible by a beginner?)

No customization needed, you just need to populate nodes in the correct way

> (i.e. would this be possible by a beginner?)

It's basic array operation after user click. And also you need to create a custom type of node with an internal state to catch click event to new nodes.

**Reply by hchaudhary1** [reply]
Hi... isn't that just a sub-flow? 

I'm trying to create a "sub-canvas". (Or perhaps I didn't fully understand your response.)

The idea would be to open a brand new clean canvas for the children.

It seems a bit complex

**Reply by ozgurrgul** [reply, marked as answer]
Yes, sub-canvas functionality you can achieve via custom nodes using subflow approach, right? 
Just render whatever complex you want only when user `enabled` them etc through a click in your parent node.
I think you can achieve to visual pretty heavy flows. Take a look at `extent: "parent" and parentNode: '' ` properties on the node object.

For example in my project, I have rendered multiple nested sub flows and there is even a dataTable and a form and it still works pretty well.

<img width="1265" alt="image" src="https://github.com/xyflow/xyflow/assets/20097288/6014c7cc-5e87-41f4-8c0e-fb0dd15b4934">

**Reply by iGmainC** [reply]
> Yes, sub-canvas functionality you can achieve via custom nodes using subflow approach, right? Just render whatever complex you want only when user `enabled` them etc through a click in your parent node. I think you can achieve to visual pretty heavy flows. Take a look at `extent: "parent" and parentNode: '' ` properties on the node object.
> 
> For example in my project, I have rendered multiple nested sub flows and there is even a dataTable and a form and it still works pretty well.
> <img alt="image" width="1265" src="https://private-user-images.githubusercontent.com/20097288/307966813-6014c7cc-5e87-41f4-8c0e-fb0dd15b4934.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2NTA1ODcsIm5iZiI6MTcyOTY1MDI4NywicGF0aCI6Ii8yMDA5NzI4OC8zMDc5NjY4MTMtNjAxNGM3Y2MtNWU4Ny00MWY0LThjMGUtZmIwZGQxNWI0OTM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIzVDAyMjQ0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM3NjZhMjgwNjU0ODAyYmY5YzRiNjNlNzdjZDI5NzVmMGUzYTFkZmI5M2IzY2QyMzg4MTZiOGQ0Y2VjMTY1MzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gAIVRCA0mRcjASXCN9u6SC7Opy0XJlBmnZLyws9ZceE">

However, using sub-flow cannot achieve the following effect:

![图片](https://github.com/user-attachments/assets/af31b54e-2663-45e1-b20c-b1bac3a8c6a0)

If the following effect is to be achieved using sub-flow, it would result in the activity range of the child nodes being able to extend over the title.

**Reply by hchaudhary1** [reply]
I am trying to create a "folder". Which can be opened, and then a new sub-canvas appears

**Comment by iGmainC** [comment]
I have a demo, but it seems to only work when the zoom level is set to 1. The zoom level of the outer canvas affects the inner canvas.


https://github.com/user-attachments/assets/293532b8-46f4-4f50-a3f9-9e0de3e181eb

**Reply by PythonLawrence** [reply]
👍 I need this functionality too.

**Comment by Sharmil001** [comment]
<img width="647" height="650" alt="Screenshot 2025-09-16 at 10 46 56 AM" src="https://github.com/user-attachments/assets/c430b78b-7a19-48fa-a7db-e299af51c83e" />

<img width="406" height="778" alt="Screenshot 2025-09-16 at 11 06 48 AM" src="https://github.com/user-attachments/assets/df90d56b-e8b7-4c97-a498-50ef25ab309f" />

**@hchaudhary1 and @ozgurrgul ozgurrgul How to Bend Parallel Edges in a Mirrored Fashion Using React Flow and Dagre Layout? How to fix that issue?**

I want to konw how you guys are managing proper mirrored edges??

//layouting code

import { span } from "@f/instrument";
import { buildNestedGroups } from "@f/utils/workflowUtils";
import dagre from "dagre";
import { Position, type Edge, type Node } from "reactflow";

export const getLayoutedElements = (
	nodes: Node[],
	edges: Edge[],
	direction: "TB" | "LR" | "RL" | "BT" = "TB",
) => {
	using console = span("getLayoutedElements");
	const dagreGraph = new dagre.graphlib.Graph();
	dagreGraph.setDefaultEdgeLabel(() => ({}));

	dagreGraph.setGraph({
		rankdir: direction,
		nodesep: 250,
		ranksep: 50,
		marginx: 25,
		marginy: 25,
	});

	const groupedNodes = new Map<string, Node[]>();

	for (const node of nodes) {
		const width = Number(node.data?.width || 150);
		const height = Number(node.data?.height || 50);

		if (width <= 0 || height <= 0) {
			console.warn(
				`Skipping node "${node.id}" due to invalid dimensions (width: ${width}, height: ${height})`,
			);
			continue;
		}

		if (node.data?.groupId) {
			if (!groupedNodes.has(node.data.groupId)) {
				groupedNodes.set(node.data.groupId, []);
			}
			groupedNodes.get(node.data.groupId)?.push(node);
		}
		try {
			if (node.type !== "group") {
				dagreGraph.setNode(node.id, { width, height });
			}
		} catch (error) {
			console.error(`Failed to add node "${node.id}" to Dagre graph:`, error);
		}
	}
	console.log("Finished adding nodes to graph");

	for (const edge of edges) {
		const sourceExists = nodes.some((node) => node.id === edge.source);
		const targetExists = nodes.some((node) => node.id === edge.target);

		if (!sourceExists || !targetExists) {
			console.warn(
				`Skipping edge from "${edge.source}" to "${edge.target}" due to missing nodes`,
			);
			continue;
		}

		try {
			dagreGraph.setEdge(edge.source, edge.target, { minlen: 1 });
		} catch (error) {
			console.error(
				`Failed to add edge from "${edge.source}" to "${edge.target}":`,
				error,
			);
		}
	}
	console.log("Finished adding edges to graph");

	try {
		dagre.layout(dagreGraph);
		console.log("Graph layout complete");
	} catch (error) {
		console.error("Error during layout calculation:", error);
		throw error;
	}

	const layoutedNodes = nodes.map((node) => {
		const nodeWithPosition = dagreGraph.node(node.id);

		if (!nodeWithPosition) {
			console.warn(`No position calculated for node "${node.id}"`);
			return node;
		}

		return {
			...node,
			targetPosition: Position.Top,
			sourcePosition: Position.Bottom,
			position: {
				x: nodeWithPosition.x - (node.data?.width || 150) / 2,
				y: nodeWithPosition.y - (node.data?.height || 50) / 2,
			},
			data: {
				...node.data,
				width: node.data?.width || 150,
				height: node.data?.height || 50,
			},
		};
	});
	console.log("Positioned layouted nodes", {
		layoutedNodeCount: layoutedNodes.length,
	});

	const groupNodes: Node[] = buildNestedGroups(nodes, dagreGraph);
	console.log("Built group nodes", { groupNodeCount: groupNodes.length });

	const resultNodes = [
		...groupNodes,
		...layoutedNodes.filter((node) => node.type !== "group"),
	];
	console.log("Layouted result prepared", {
		finalNodeCount: resultNodes.length,
		finalEdgeCount: edges.length,
	});

	return {
		nodes: resultNodes,
		edges,
	};
};


//buildnested groups

export const buildNestedGroups = (nodes: Node[], dagreGraph: any) => {
	const processedGroups = new Set<string>();
	const groupNodes: Node[] = [];
	const nestedGroups = new Map<string, string>();

	const getGroupNodes = (groupId: string) =>
		nodes.filter((node) => node.data?.groupId === groupId);

	for (const node of nodes) {
		if (node.type === "group" && node.data.groupId) {
			if (!nestedGroups.has(node.id)) {
				nestedGroups.set(node.id, node.data.groupId);
			}
		}
	}

	const createGroup = (groupId: string, parentGroupId: string | undefined) => {
		if (processedGroups.has(groupId)) return;
		const group = getGroupNodes(groupId);
		if (group.length === 0) return;
		let minX = Number.POSITIVE_INFINITY;
		let maxX = Number.NEGATIVE_INFINITY;
		let minY = Number.POSITIVE_INFINITY;
		let maxY = Number.NEGATIVE_INFINITY;

		for (const node of group) {
			const n = dagreGraph.node(node.id);
			if (!n) continue;

			const nodeWidth = node.data?.width || 150;
			const nodeHeight = node.data?.height || 50;

			const nodeLeft = n.x - nodeWidth / 2;
			const nodeRight = n.x + nodeWidth / 2;
			const nodeTop = n.y - nodeHeight / 2;
			const nodeBottom = n.y + nodeHeight / 2;

			minX = Math.min(minX, nodeLeft);
			maxX = Math.max(maxX, nodeRight);
			minY = Math.min(minY, nodeTop);
			maxY = Math.max(maxY, nodeBottom);
		}

		if (
			minX === Number.POSITIVE_INFINITY ||
			minY === Number.POSITIVE_INFINITY ||
			maxX === Number.NEGATIVE_INFINITY ||
			maxY === Number.NEGATIVE_INFINITY
		) {
			console.warn(`Skipping group ${groupId} due to invalid bounds`);
			return;
		}

		const basePadding = 20;

		const bgWidth = Math.max(0, maxX - minX + 2 * basePadding);
		const bgHeight = Math.max(0, maxY - minY + 2 * basePadding);

		groupNodes.push({
			id: `${groupId}`,
			type: "group",
			position: { x: minX - basePadding, y: minY - basePadding },
			data: { label: `Loop: ${groupId}`, groupId: parentGroupId },
			style: {
				width: bgWidth,
				height: bgHeight,
				border: "2px dashed var(--utility-gray-300)",
				backgroundColor: "var(--utility-gray-50)",
				borderRadius: "28px",
				zIndex: -1,
			},
		});

		processedGroups.add(groupId);
	};

	const uniqueGroupIds = new Set(
		nodes.flatMap((node) => (node.data?.groupId ? [node.data.groupId] : [])),
	);

	for (const groupId of uniqueGroupIds) {
		const parentGroupId = nestedGroups.has(groupId)
			? nestedGroups.get(groupId)
			: undefined;
		createGroup(groupId, parentGroupId);
	}

	uniqueGroupIds.forEach(createGroup);

	const updateParentWidths = (childId: string, widthIncrease: number) => {
		using console = span("updateParentWidths");
		let currentId: string | undefined = childId;
		console.log("Updating parent widths for childId:", childId);

		while (currentId) {
			const parentId = nestedGroups.get(currentId);
			if (!parentId) break;

			const parentGroup = groupNodes.find((g) => g.id === `${parentId}`);
			const childGroup = groupNodes.find((g) => g.id === `${currentId}`);

			if (parentGroup && childGroup && parentGroup.style && childGroup.style) {
				const newWidth = Number(childGroup.style.width) + widthIncrease;

				if (newWidth > Number(parentGroup.style.width)) {
					parentGroup.style.width = newWidth;
					parentGroup.position.x = Math.min(
						parentGroup.position.x,
						childGroup.position.x - widthIncrease / 2,
					);
				}
			}
			currentId = parentId;
		}
	};

	for (const [childId, _] of nestedGroups) {
		updateParentWidths(childId, 60);
	}

	console.log("Finalized updateParentWidths", {
		groupNodes,
	});

	return groupNodes;
};
